### PR TITLE
x86 fp16 weight storage optimizations

### PR DESCRIPTION
**Pull request by Timen**
I wanted to add the ability to store weights as fp16 and convert to fp32 when using for arithmetic. Main advantage of this is lowering the cache pressure, resulting in a small speedup for most layer and a significant speedup for the innerproduct layer. 

Major changes:
1. added fp16 weight storage option
2. support fp16 weight storage in conv 1x1,3x3 and dwconv 3,3
3. support fp16 weight storage in innerproduct when using elempack 8. 

I'm open to suggestions about fp16 weight storage, I know bfloat16 is being worked on in the arm side, but I haven't found a method fast enough to support bfloat16 featuremaps. 

Another question, @nihui is it okay if I add back an LSTM version that stores the state between network executions? This was removed a while back but I'm not sure why, I would also like to add fp16 weight storage to an x86 LSTM version.

### CODE REVIEW

**Comment by nihui** [comment]
The last time I tested on the arm, I found that the fp16-fp32 conversion cost is higher than direct fp32 read and write. The conversion of bfloat16-fp32 only needs to be realized by shift, which is very fast, so it has an acceleration effect on some low-end arm-cpu.

I don't know if x86 is the same. Since weight can be accelerated with fp16 storage, what if blob also uses fp16 storage?

**Comment by nihui** [comment]
The Option class already has use_fp16_storage option, no need to add a new one.

LSTM is now changed to the same as defined in the onnx operator documentation, but does not implement the input and output of hidden cells and other intermediate states. If you want to achieve this, please refer to the definition in the onnx operator documentation, maybe you can use more paramdict parameters to maintain behavioral compatibility.

**Comment by tijmen-loomai** [comment]
Fp16 to fp32 conversion is quite fast on x86 with the fp16c vector extension instructions. At least fast enough for loading and storing weights while also reducing cache pressure and ram usage. What is your experience with regards to accuracy when using bfloat16 to execute fp32 models? Does it handle it well or does a model specifically need to be trained with bfloat16 for it to make sense?

I’m not sure if fp16 is the right way to go for featuremaps, mostly because i’m not sure there is a fast way to broadcast individual elements of an avx register. I will need to do some tests, I also haven’t found a simple way to do bfloat16 to fp32 conversion using avx intrinsics.

**Comment by Timen** [comment]
Fp16 to fp32 conversion is quite fast on x86 with the fp16c vector extension instructions. At least fast enough for loading and storing weights while also reducing cache pressure and ram usage. What is your experience with regards to accuracy when using bfloat16 to execute fp32 models? Does it handle it well or does a model specifically need to be trained with bfloat16 for it to make sense?

I’m not sure if fp16 is the right way to go for featuremaps, mostly because i’m not sure there is a fast way to broadcast individual elements of an avx register. I will need to do some tests, I also haven’t found a simple way to do bfloat16 to fp32 conversion using avx intrinsics.

**Comment by tencent-adm** [comment]
[CLA assistant check](https://cla.opensource.tencent.com/Tencent/ncnn?pullRequest=1871) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.opensource.tencent.com/Tencent/ncnn?pullRequest=1871) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: Timen<br/>:x: restyled-commits<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.tencent.com/check/Tencent/ncnn?pullRequest=1871) it.</sub>

**Comment by Timen** [comment]
Ready for review 👍  I think I will do some fp16 blob storage in a next pr

File: src/layer/x86/convolution_3x3_pack8_fp16.h [hunk 1]
```diff
@@ -0,0 +1,1487 @@
...
+                {
+                    kernel_tm0[j * 8 + i] = tmpp[0] * ktm[i][0] + tmpp[1] * ktm[i][1] + tmpp[2] * ktm[i][2];
+                }
+            }
+        }
+    }
+    // interleave
+    // src = 64-inch-outch
+    // dst = 8b-8a-inch/8a-84-outch/8b;
+    kernel_tm_pack8.create(inch / 8, 64, (outch / 4) / 2 + (outch / 4) % 2, (size_t)2u * 64, 64);
```
Comment (nihui) [inline src/layer/x86/convolution_3x3_pack8_fp16.h:69]: outch must be a multiple of 8 here
```(outch / 4) / 2 + (outch / 4) % 2``` to ```outch / 8```

File: src/layer/x86/convolution_3x3_pack8.h [hunk 2]
```diff
@@ -76,7 +66,7 @@ static void conv3x3s1_winograd64_transform_kernel_pack8_avx(const Mat& kernel, M
     // interleave
     // src = 64-inch-outch
     // dst = 4b-4a-inch/4a-64-outch/4b;
-    kernel_tm_pack8.create(2 * inch / 8, 64, (outch / 4) / 2 + (outch / 4) % 2, (size_t)4u * 64, 64);
+    kernel_tm_pack8.create(inch / 8, 64, (outch / 4) / 2 + (outch / 4) % 2, (size_t)4u * 64, 64);
```
Comment (nihui) [inline src/layer/x86/convolution_3x3_pack8.h:69]: outch must be a multiple of 8 here
(outch / 4) / 2 + (outch / 4) % 2 to outch / 8

File: src/layer/x86/avx_usability.h [hunk 3]
```diff
@@ -18,6 +18,19 @@
 #define AVX_USABILITY
 #include <immintrin.h>
 
+static inline __m256 loadfp16(const unsigned short* ptr)
+{
+    return _mm256_cvtph_ps(_mm_load_si128((__m128i*)(ptr)));
```
[the remark this answers is not in the document]
Comment (tijmen-loomai) [reply src/layer/x86/avx_usability.h:23]: Done! (switched to: _mm_lddqu_si128)

File: src/layer/x86/innerproduct_x86.cpp [hunk 4]
```diff
@@ -1,223 +1,588 @@
...
+    return 0;
+#else
+    return InnerProduct::forward(bottom_blob, top_blob, opt);
+#endif // __AVX__
+}
+#if __AVX__
+
+int InnerProduct_x86::forward_fp16(const Mat& bottom_blob, Mat& top_blob,
+                                   const Option& opt) const
+{
```
[the remark this answers is not in the document]
Comment (tijmen-loomai) [reply src/layer/x86/innerproduct_x86.cpp:376]: Okay I added the remain function, had to fill an 8 lane vector before converting from fp16 to fp32 as I'm not aware of a quick way to do this per element.

File: src/layer/x86/convolution_x86.cpp [hunk 5]
```diff
@@ -589,24 +622,61 @@ int Convolution_x86::forward(const Mat& bottom_blob, Mat& top_blob, const Option
...
+            else
+            {
+                conv1x1s2_pack8_avx(bottom_blob_bordered, top_blob, weight_data_pack8, bias_data, opt);
+            }
+
             if (activation)
             {
                 activation->forward_inplace(top_blob, opt);
             }
         }
```
[the remark this answers is not in the document]
Comment (tijmen-loomai) [reply src/layer/x86/convolution_x86.cpp:653]: removed

File: src/layer/x86/convolution_x86.cpp [hunk 6]
```diff
@@ -188,188 +191,211 @@ int Convolution_x86::create_pipeline(const Option& opt)
...
         else
         {
-            // src = kw-kh-inch-outch
-            // dst = 8b-8a-kw-kh-inch/8a-outch/8b
-            Mat weight_data_r2 = weight_data.reshape(maxk, num_input, num_output);
-
-            weight_data_pack8.create(maxk, num_input / 8, num_output / 8, (size_t)4 * 64, 64);
-
-            for (int q = 0; q + 7 < num_output; q += 8)
+            if (kernel_w == 3 && kernel_h == 3 && dilation_w == 1 && dilation_h == 1 && stride_w == 1 && stride_h == 1)
```
[the remark this answers is not in the document]
Comment (tijmen-loomai) [reply src/layer/x86/convolution_x86.cpp:212]: done

File: src/layer/x86/convolution_3x3_pack8_fp16.h (hunk 1 above)
Comment (tijmen-loomai) [reply src/layer/x86/convolution_3x3_pack8_fp16.h:69]: Done I settled on:    `kernel_tm_pack8.create(inch / 8, 64, outch / 8, (size_t)4u * 64, 64);`
which seems to work :)

File: src/layer/x86/convolution_3x3_pack8.h (hunk 2 above)
Comment (tijmen-loomai) [reply src/layer/x86/convolution_3x3_pack8.h:69]: see above

File: src/layer/x86/innerproduct_x86.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Timen) [reply src/layer/x86/innerproduct_x86.cpp:376]: Okay I added the remain function, had to fill an 8 lane vector before converting from fp16 to fp32 as I'm not aware of a quick way to do this per element.

File: src/layer/x86/convolution_3x3_pack8_fp16.h (hunk 1 above)
Comment (Timen) [reply src/layer/x86/convolution_3x3_pack8_fp16.h]: I settled on: `kernel_tm_pack8.create(inch / 8, 64, outch / 8, (size_t)4u * 64, 64);` your suggestion caused a segfault and this makes sense give the 8 lane memory layour.

File: src/layer/x86/avx_usability.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (Timen) [reply src/layer/x86/avx_usability.h]: switched to: _mm_lddqu_si128

File: src/layer/x86/convolution_3x3_pack8_fp16.h (hunk 1 above)
Comment (nihui) [reply src/layer/x86/convolution_3x3_pack8_fp16.h]: > I settled on: `kernel_tm_pack8.create(inch / 8, 64, outch / 8, (size_t)4u * 64, 64);` your suggestion caused a segfault and this makes sense give the 8 lane memory layour.

yes, that's what I mean. outch / 8 for the channel dimension size

**Comment by nihui** [comment]
:+1:

**Comment by nihui** [comment]
Look forward to the effect of using fp16 storage on feature maps :smile:

**Pull request merged**